### PR TITLE
Document file data sources

### PR DESCRIPTION
**Pull request by leoyvens**
Taken from this draft by @azf20 https://hackmd.io/EOG_Cw_4SIGR1D2qyS1Dyg.

### CODE REVIEW

**Review by schmidsi — COMMENTED** [review]
Some remarks from trying to follow the docs. @MichaelMacaulay might be able to incorporate this feedback.

File: pages/en/developing/creating-a-subgraph.mdx [hunk 1]
```diff
@@ -956,3 +956,220 @@ The grafted subgraph can use a GraphQL schema that is not identical to the one o
...
+import { TokenMetadata as TokenMetadataTemplate } from '../generated/templates'
+
+export function handleTransfer(event: TransferEvent): void {
+  let token = Token.load(event.params.tokenId.toString())
+  if (!token) {
+    token = new Token(event.params.tokenId.toString())
+    token.tokenID = event.params.tokenId
+
+    token.tokenURI = '/' + event.params.tokenId.toString() + '.json'
+    const tokenIpfsHash = ipfshash + token.tokenURI
```
[the remark this answers is not in the document]
Comment (Web3Slimchance) [reply pages/en/developing/creating-a-subgraph.mdx:1111]: Added the IPFS hash and two explaining comments

File: pages/en/developing/creating-a-subgraph.mdx [hunk 2]
```diff
@@ -956,3 +956,220 @@ The grafted subgraph can use a GraphQL schema that is not identical to the one o
...
+The CID of the file as a readable string can be accessed via the `dataSource` as follows:
+
+```typescript
+const cid = dataSource.stringParam()
+```
+
+Example handler:
+
+```typescript
+import { json, Bytes, dataSource } from '@graphprotocol/graph-ts'
```
[the remark this answers is not in the document]
Comment (Web3Slimchance) [reply pages/en/developing/creating-a-subgraph.mdx:1067]: Added

File: pages/en/developing/creating-a-subgraph.mdx [hunk 3]
```diff
@@ -956,3 +956,220 @@ The grafted subgraph can use a GraphQL schema that is not identical to the one o
...
+  owner: User!
+}
+
+type TokenMetadata @entity {
+  id: ID!
+  image: String!
+  externalURL: String!
+  name: String!
+  description: String!
+  type: String!
```
[the remark this answers is not in the document]
Comment (Web3Slimchance) [reply pages/en/developing/creating-a-subgraph.mdx:1022]: Removed 👍

File: pages/en/developing/creating-a-subgraph.mdx [hunk 4]
```diff
@@ -956,3 +956,219 @@ The grafted subgraph can use a GraphQL schema that is not identical to the one o
...
+
+> Currently Graph Node supports [v0 and v1 content identifiers](https://docs.ipfs.tech/concepts/content-addressing/), and content identifers with directories (e.g. `bafyreighykzv2we26wfrbzkcdw37sbrby4upq7ae3aqobbq7i4er3tnxci/metadata.json`)
+
+Example:
+
+```typescript
+import { TokenMetadata as TokenMetadataTemplate } from '../generated/templates'
+
+const ipfshash = "QmaXzZhcYnsisuue5WRdQDH6FDvqkLQX1NckLqBYeYYEfm"
+//This example code is for a Crypto coven subgraph. The above ipfs hash is a directory with token metadata for all crypto coven NFTs.
```
Comment (schmidsi) [inline pages/en/developing/creating-a-subgraph.mdx:1105]: Perfect!

**Review by schmidsi — APPROVED** [review]

**Pull request merged**